src/managers/upgrade.py

```diff
@@ -0,0 +1,46 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Manager for building necessary files for TLS auth."""
+import logging
+
+from charms.data_platform_libs.v0.upgrade import DependencyModel
+
+from core.cluster import SUBSTRATES, ClusterState
+from core.workload import WorkloadBase
+
+logger = logging.getLogger(__name__)
+
+
+class UpgradeManager:
+    """Manager for building necessary files for Java TLS auth."""
+
+    def __init__(
+        self,
+        state: ClusterState,
+        workload: WorkloadBase,
+        substrate: SUBSTRATES,
+        dependency_model: DependencyModel,
+    ):
+        self.state = state
+        self.workload = workload
+        self.substrate = substrate
+        self.dependency_model = dependency_model
+
+    def version_compatible(self) -> bool:
+        """Verify version compatibility with Opensearch."""
+        # When there's no Opensearch connection, we shouldn't report version mismatch
+        if not self.state.opensearch_server:
+            return True
```

Comment:
1. Linting won't let us get rid of this -- as in the next line I'm referring to a `self.state.opensearch_server` attribute.
2. We consider that there is no verison compatibility issue when there is no connection to Opensearch. To me this fully makes sense (like mulitplication with zero results in zero). It's a base definition/assumption.

I'm not sure if I got your argumetnation correctly. 

> if the server is down

Do you mean the Openserach server? The above check is independent from the Openserach server being up or down. The above check is stricly looking at the existence of an Opensearch relation.

Reply:
See https://github.com/canonical/opensearch-dashboards-operator/pull/91/files/36eba6f1549b584f11ec960840e8c3500407d5b7#diff-9dc137d1679e723fa586c453222bc0a25cfa1dccf43430336148c13fd177c382L46

I think we should keep the scope of our PRs somewhat consize, not to change an upgrade strategy in a "Healthcheck/Status"-related modification.

Reply:
1. It is fine. We can exclude this warning if needed.
2. Should the dashboards charm not be able to upgrade unless related to OpenSearch?

Reply:
Dashboards charm is `blocked` (as a 'charm') if not related to Opensearch.
As a 'service' it's literally unavailable. The service doesn't come up.

I assume in real life most OSD Rolling Upgrades would happen
 - when the charm is in an `active` state
 - the service is healthy
 
Yet.

The current checks also allow for OSD upgrades when not related to Opensearch. Technically possible. After all, it's a valid scenario (as long as upgrades libs allow for an upgrade from a `blocked` status).

In this case potential version incompatibility will be detected during the relation establishing (as it will trigger `reconcile()` right after the reboot, and new healthchecks will detect the mis-match.).